src/node_wasi.cc

```diff
@@ -0,0 +1,1837 @@
+  uvwasi_errno_t err = uvwasi_sock_shutdown(&wasi->uvw_, sock, how);
+  args.GetReturnValue().Set(err);
+}
+
+
+void WASI::_SetMemory(const FunctionCallbackInfo<Value>& args) {
+  WASI* wasi;
+  CHECK_EQ(args.Length(), 1);
+  CHECK(args[0]->IsObject());
+  ASSIGN_OR_RETURN_UNWRAP(&wasi, args.This());
+  wasi->memory_.Reset(wasi->env()->isolate(), args[0].As<Object>());
+}
+
+
+void WASI::readUInt8(char* memory, uint8_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = memory[offset] & 0xFF;
+}
+
+
+void WASI::readUInt16(char* memory, uint16_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = (memory[offset] & 0xFF) |
+           ((memory[offset + 1] & 0xFF) << 8);
+}
+
+
+void WASI::readUInt32(char* memory, uint32_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  *value = (memory[offset] & 0xFF) |
+           ((memory[offset + 1] & 0xFF) << 8) |
+           ((memory[offset + 2] & 0xFF) << 16) |
+           ((memory[offset + 3] & 0xFF) << 24);
+}
+
+
+void WASI::readUInt64(char* memory, uint64_t* value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  CHECK_NOT_NULL(value);
+  uint64_t low = (memory[offset] & 0xFF) |
+                 ((memory[offset + 1] & 0xFF) << 8) |
+                 ((memory[offset + 2] & 0xFF) << 16) |
+                 ((memory[offset + 3] & 0xFF) << 24);
+  uint64_t high = (memory[offset + 4] & 0xFF) |
+                  ((memory[offset + 5] & 0xFF) << 8) |
+                  ((memory[offset + 6] & 0xFF) << 16) |
+                  ((memory[offset + 7] & 0xFF) << 24);
+  *value = (high << 32) + low;
+}
+
+
+void WASI::writeUInt8(char* memory, uint8_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset] = value & 0xFF;
+}
+
+
+void WASI::writeUInt16(char* memory, uint16_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset] = (value >> 8) & 0xFF;
+}
+
+
+void WASI::writeUInt32(char* memory, uint32_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset++] = (value >> 8) & 0xFF;
+  memory[offset++] = (value >> 16) & 0xFF;
+  memory[offset] = (value >> 24) & 0xFF;
+}
+
+
+void WASI::writeUInt64(char* memory, uint64_t value, uint32_t offset) {
+  CHECK_NOT_NULL(memory);
+  memory[offset++] = value & 0xFF;
+  memory[offset++] = (value >> 8) & 0xFF;
+  memory[offset++] = (value >> 16) & 0xFF;
+  memory[offset++] = (value >> 24) & 0xFF;
+  memory[offset++] = (value >> 32) & 0xFF;
+  memory[offset++] = (value >> 40) & 0xFF;
+  memory[offset++] = (value >> 48) & 0xFF;
+  memory[offset] = (value >> 56) & 0xFF;
+}
+
+
+uvwasi_errno_t WASI::backingStore(char** store, size_t* byte_length) {
+  Environment* env = this->env();
+  Local<Object> memory = PersistentToLocal::Default(env->isolate(),
+                                                    this->memory_);
```

Comment:
If `memory_` is always a strong reference, feel free to use `PersistentToLocal::Strong`.